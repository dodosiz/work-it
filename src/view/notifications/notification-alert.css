.notification-alert {
	position: fixed;
	bottom: 1em;
	right: 1em;
	transition: all 250ms linear;
	width: 40%;
}
