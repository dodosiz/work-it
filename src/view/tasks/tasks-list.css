.task {
	box-shadow: rgba(0, 0, 0, 0.24) 0px 3px 8px;
	margin: 2em 0;
	opacity: 100%;
	transition: opacity 2s;
}

.pointer {
	cursor: pointer;
}

.done {
	color: green;
}

.undone {
	color: lightgreen;
}
