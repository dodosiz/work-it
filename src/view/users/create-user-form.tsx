import * as React from "react";
import { Button, Form } from "react-bootstrap";
import { useDispatch } from "react-redux";
import { addUser, closeUserForm } from "../../data/users";

export function CreateUserForm() {
	const dispatch = useDispatch();
	const [firstName, setFirstName] = React.useState("");
	const [lastName, setLastName] = React.useState("");
	const [role, setRole] = React.useState("");
	const handleFirstNameChange: React.ChangeEventHandler<HTMLInputElement> = (
		event
	) => {
		setFirstName(event.target.value);
	};
	const handleLastNameChange: React.ChangeEventHandler<HTMLInputElement> = (
		event
	) => {
		setLastName(event.target.value);
	};
	const handleRoleChange: React.ChangeEventHandler<HTMLInputElement> = (
		event
	) => {
		setRole(event.target.value);
	};
	const handleSubmit: React.FormEventHandler<HTMLFormElement> = (event) => {
		event.preventDefault();
		dispatch(addUser({ firstName, lastName, role }));
		dispatch(closeUserForm());
	};
	const submitDisabled = () => {
		return !(firstName.length && lastName.length && role.length);
	};
	return (
		<Form onSubmit={handleSubmit}>
			<Form.Group className="mb-3" controlId="firstName">
				<Form.Label>First name:</Form.Label>
				<Form.Control
					type="text"
					value={firstName}
					onChange={handleFirstNameChange}
				/>
			</Form.Group>
			<Form.Group className="mb-3" controlId="lastName">
				<Form.Label>Last name:</Form.Label>
				<Form.Control
					type="text"
					value={lastName}
					onChange={handleLastNameChange}
				/>
			</Form.Group>
			<Form.Group className="mb-3" controlId="role">
				<Form.Label>Role:</Form.Label>
				<Form.Control type="text" value={role} onChange={handleRoleChange} />
			</Form.Group>
			<Button disabled={submitDisabled()} variant="primary" type="submit">
				Submit
			</Button>
		</Form>
	);
}
