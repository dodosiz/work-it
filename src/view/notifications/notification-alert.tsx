import * as React from "react";
import { Alert, Fade } from "react-bootstrap";
import { useDispatch, useSelector } from "react-redux";
import {
	clearNotification,
	notificationSelector,
} from "../../data/notifications/notifications";
import "./notification-alert.css";

export function NotificationAlert() {
	const notificationTimeout = 2000;
	const dispatch = useDispatch();
	const notification = useSelector(notificationSelector);
	const show = !!notification;
	const fadeOut = () => {
		setTimeout(() => {
			dispatch(clearNotification());
		}, notificationTimeout);
	};
	const handleClose = () => {
		dispatch(clearNotification());
	};
	return (
		<Fade in={show} timeout={notificationTimeout} onEntered={() => fadeOut()}>
			<Alert
				className="notification-alert"
				dismissible
				show={show}
				variant="success"
				onClose={() => handleClose()}
			>
				{notification}
			</Alert>
		</Fade>
	);
}
