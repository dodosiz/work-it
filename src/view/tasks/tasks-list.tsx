import * as React from "react";
import { Button, Card, Col, Row } from "react-bootstrap";
import {
	BsCheckCircle,
	BsCheckCircleFill,
	BsFillTrashFill,
	BsPencilFill,
} from "react-icons/bs";
import { useDispatch, useSelector } from "react-redux";
import { addNotification } from "../../data/notifications/notifications";
import {
	deleteTask,
	finishTask,
	openTaskForm,
	Task,
	tasksSelector,
} from "../../data/tasks/tasks";
import "./tasks-list.css";

export function TasksList(props: { mode: "todo" | "done" }) {
	const tasks = useSelector(tasksSelector(props.mode));
	if (tasks.length === 0) {
		return (
			<p>
				{props.mode === "todo" ? "No tasks found." : "No finished tasks found."}
			</p>
		);
	}
	return (
		<>
			{tasks.sort(sortByDate).map((task) => (
				<TaskCard key={task.id} task={task} mode={props.mode} />
			))}
		</>
	);
}

function sortByDate(task1: Task, task2: Task) {
	if (task1.dateFinished && task2.dateFinished) {
		return task2.dateFinished.localeCompare(task1.dateFinished);
	} else {
		return 0;
	}
}

interface TaskCardProps {
	mode: "todo" | "done";
	task: Task;
}

function TaskCard(props: TaskCardProps) {
	const fadeOutTimeout = 1500; // a little faster then the CSS transition
	const [checked, setChecked] = React.useState(false);
	const dispatch = useDispatch();
	const handleCheck = () => {
		setChecked(true);
		setTimeout(() => {
			dispatch(finishTask({ taskId: props.task.id }));
		}, fadeOutTimeout);
	};
	const handleEdit = () => {
		dispatch(
			openTaskForm({
				taskFormMode: "edit",
				clickedTask: props.task,
			})
		);
	};
	const handleDelete = () => {
		dispatch(deleteTask({ taskId: props.task.id }));
		dispatch(
			addNotification({
				message: `Deleted task "${props.task.title}".`,
			})
		);
	};
	const handleTitleClick = () => {
		dispatch(
			openTaskForm({
				taskFormMode: "readonly",
				clickedTask: props.task,
			})
		);
	};
	return (
		<Card className="task" style={checked ? { opacity: "0%" } : {}}>
			<Card.Body>
				<Row>
					<Col md={1}>
						<CheckBox
							taskId={props.task.id}
							checked={props.mode === "todo" ? checked : true}
							handleCheck={props.mode === "todo" ? handleCheck : undefined}
						/>
					</Col>
					<Col md={5}>
						<span
							className="pointer"
							data-testid={`title-${props.task.id}`}
							onClick={handleTitleClick}
						>
							{props.task.title}
						</span>
					</Col>
					<Col md={4}>
						{props.task.assignee?.firstName +
							" " +
							props.task.assignee?.lastName}
					</Col>
					{props.mode === "done" && (
						<Col md={2}>{`Finished on: ${props.task.dateFinished}`}</Col>
					)}
					{props.mode === "todo" && (
						<Col md={2}>
							<Button
								data-testid={`edit-button-${props.task.id}`}
								variant="outline-primary"
								className="me-2"
								onClick={handleEdit}
							>
								<BsPencilFill />
							</Button>
							<Button
								data-testid={`delete-button-${props.task.id}`}
								variant="outline-danger"
								onClick={handleDelete}
							>
								<BsFillTrashFill />
							</Button>
						</Col>
					)}
				</Row>
			</Card.Body>
		</Card>
	);
}

interface CheckBoxProps {
	taskId: string;
	checked: boolean;
	handleCheck?: () => void;
}

function CheckBox(props: CheckBoxProps) {
	return (
		<div
			className={`${props.checked ? "done" : "undone"}`}
			data-testid={`check-${props.taskId}`}
		>
			{!props.checked && (
				<BsCheckCircle
					className="pointer"
					onClick={props.handleCheck}
					size={"1.5em"}
					data-testid={`check-box-${props.taskId}`}
				/>
			)}
			{props.checked && <BsCheckCircleFill size={"1.5em"} />}
		</div>
	);
}
