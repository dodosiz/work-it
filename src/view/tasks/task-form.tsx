import * as React from "react";
import { Button, Form } from "react-bootstrap";
import { useDispatch, useSelector } from "react-redux";
import { addNotification } from "../../data/notifications/notifications";
import {
	clickedTaskSelector,
	closeTaskForm,
	createTask,
	taskFormModeSelector,
} from "../../data/tasks/tasks";
import { usersSelector } from "../../data/users/users";

export function TaskForm() {
	const dispatch = useDispatch();
	const users = useSelector(usersSelector);
	const clickedTask = useSelector(clickedTaskSelector);
	const readOnly = useSelector(taskFormModeSelector) === "readonly";
	const [title, setTitle] = React.useState(
		clickedTask ? clickedTask.title : ""
	);
	const [description, setDescription] = React.useState(
		clickedTask ? clickedTask?.description : ""
	);
	const [userId, setUserId] = React.useState<string | undefined>(
		clickedTask ? clickedTask.assignee?.id : undefined
	);
	const [dirty, setDirty] = React.useState(false);
	const handleTitleChange: React.ChangeEventHandler<HTMLInputElement> = (
		event
	) => {
		setTitle(event.target.value);
		setDirty(true);
	};
	const handleDescriptionChange: React.ChangeEventHandler<HTMLInputElement> = (
		event
	) => {
		setDescription(event.target.value);
		setDirty(true);
	};
	const handleAssigneeChange: React.ChangeEventHandler<HTMLSelectElement> = (
		event
	) => {
		setUserId(event.target.value);
		setDirty(true);
	};
	const handleSubmit: React.FormEventHandler<HTMLFormElement> = (event) => {
		event.preventDefault();
		if (!readOnly) {
			const assignee = users.find((user) => user.id === userId);
			dispatch(createTask({ title, description, assignee }));
			dispatch(
				addNotification({
					message: `Created task: ${title}`,
				})
			);
			dispatch(closeTaskForm());
		}
	};
	const handleCancel = () => {
		dispatch(closeTaskForm());
	};
	const submitDisabled = () => {
		return !isValid() || !dirty;
	};
	const isValid = () => {
		return title && userId;
	};
	return (
		<Form onSubmit={handleSubmit}>
			<Form.Group className="mb-3" controlId="title">
				<Form.Label>Title:</Form.Label>
				<Form.Control
					readOnly={readOnly}
					type="text"
					value={title}
					onChange={handleTitleChange}
				/>
			</Form.Group>
			<Form.Group className="mb-3" controlId="description">
				<Form.Label>Description:</Form.Label>
				<Form.Control
					type="textarea"
					readOnly={readOnly}
					value={description}
					onChange={handleDescriptionChange}
				/>
			</Form.Group>
			<Form.Group className="mb-3" controlId="assignee">
				<Form.Select
					aria-label="Assignee"
					onChange={handleAssigneeChange}
					disabled={readOnly}
					value={userId}
					data-testid="assignee"
				>
					<option></option>
					{users.map((user) => (
						<option key={user.id} value={user.id}>
							{user.firstName + " " + user.lastName}
						</option>
					))}
				</Form.Select>
			</Form.Group>
			{!readOnly && (
				<>
					<Button
						className="me-2"
						disabled={submitDisabled()}
						variant="primary"
						type="submit"
					>
						Save
					</Button>
					<Button variant="secondary" onClick={() => handleCancel()}>
						Cancel
					</Button>
				</>
			)}
		</Form>
	);
}
