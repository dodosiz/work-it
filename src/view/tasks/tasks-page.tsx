import * as React from "react";
import { Container, Navbar, Nav, Button } from "react-bootstrap";
import { useDispatch, useSelector } from "react-redux";
import {
	closeTaskForm,
	openTaskForm,
	taskFormModeSelector,
	taskFormOpenedSelector,
} from "../../data/tasks/tasks";
import { TaskFilter } from "../filter/task-filter";
import { ModalDialog } from "../modal/modal-dialog";
import { TaskForm } from "./task-form";
import { TasksList } from "./tasks-list";

export function TasksPage(props: { mode: "todo" | "done" }) {
	const dispatch = useDispatch();
	const taskFormOpened = useSelector(taskFormOpenedSelector);
	const taskFormMode = useSelector(taskFormModeSelector);
	const handleCloseTaskForm = () => {
		dispatch(closeTaskForm());
	};
	const handleCreateTaskClick = () => {
		dispatch(openTaskForm({ taskFormMode: "edit" }));
	};
	return (
		<>
			<ModalDialog
				handleClose={handleCloseTaskForm}
				show={taskFormOpened}
				title={taskFormMode === "readonly" ? "View task" : "Create new task"}
			>
				<TaskForm />
			</ModalDialog>
			<Navbar expand="lg" variant="light" bg="light">
				<Container>
					<Nav className="me-auto">
						{props.mode === "todo" && (
							<Button variant="success" onClick={handleCreateTaskClick}>
								Create Task
							</Button>
						)}
					</Nav>
					<TaskFilter />
				</Container>
			</Navbar>
			<Container>
				<TasksList mode={props.mode} />
			</Container>
		</>
	);
}
