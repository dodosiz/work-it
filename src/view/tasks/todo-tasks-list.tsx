import * as React from "react";
import { Card, Col, Row } from "react-bootstrap";
import { Task } from "../../data/tasks/tasks";
import "./tasks-list.css";

export function TodoTasksList(props: { tasks: Task[] }) {
	if (props.tasks.length === 0) {
		return <p>No tasks to do. Start by creating a new task.</p>;
	}
	return (
		<>
			{props.tasks.map((task) => (
				<TodoTask key={task.id} task={task} />
			))}
		</>
	);
}

function TodoTask(props: { task: Task }) {
	return (
		<Card className="task">
			<Card.Body>
				<Row>
					<Col>{props.task.title}</Col>
					<Col>
						{props.task.assignee?.firstName +
							" " +
							props.task.assignee?.lastName}
					</Col>
				</Row>
			</Card.Body>
		</Card>
	);
}
