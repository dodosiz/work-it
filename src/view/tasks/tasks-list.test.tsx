import * as React from "react";
import { tasksReducer } from "../../data/tasks/tasks";
import { TasksList } from "./tasks-list";
import { render, screen } from "@testing-library/react";
import { configureStore, EnhancedStore } from "@reduxjs/toolkit";
import { usersReducer } from "../../data/users/users";
import { notificationsReducer } from "../../data/notifications/notifications";
import { Provider } from "react-redux";
import userEvent from "@testing-library/user-event";

describe("Tasks list:", () => {
	let store: EnhancedStore;
	beforeEach(() => {
		store = configureStore({
			reducer: {
				usersState: usersReducer,
				notifications: notificationsReducer,
				tasksState: tasksReducer,
			},
			preloadedState: {
				notifications: undefined,
				tasksState: {
					taskFormOpened: false,
					tasks: [
						{
							id: "task1",
							assignee: {
								id: "user1",
								firstName: "Alejandro",
								lastName: "Matthews",
								role: "Project manager",
							},
							title: "Plan the requirements for the next release.",
							dateFinished: undefined,
							description: "",
						},
						{
							id: "task2",
							assignee: {
								id: "user1",
								firstName: "Alejandro",
								lastName: "Matthews",
								role: "Project manager",
							},
							title:
								"Meet with the stakeholders to discuss the new requirements.",
							dateFinished: undefined,
							description: "",
						},
						{
							id: "task3",
							assignee: {
								id: "user1",
								firstName: "Alejandro",
								lastName: "Matthews",
								role: "Project manager",
							},
							title:
								"Meet with the stakeholders to discuss the new requirements.",
							dateFinished: "2020-01-01",
							description: "",
						},
					],
				},
				usersState: {
					userFormOpened: false,
					userIdToEdit: undefined,
					users: [
						{
							id: "user1",
							firstName: "Alejandro",
							lastName: "Matthews",
							role: "Project manager",
						},
					],
				},
			},
		});
	});
	test("todo list", () => {
		render(
			<Provider store={store}>
				<TasksList mode={"todo"} />
			</Provider>
		);
		expect(screen.queryAllByTestId("check-task1").length).toBe(1);
		expect(screen.queryAllByTestId("check-task2").length).toBe(1);
		expect(screen.queryAllByTestId("check-task3").length).toBe(0);
	});
	test("done list", () => {
		render(
			<Provider store={store}>
				<TasksList mode={"done"} />
			</Provider>
		);
		expect(screen.queryAllByTestId("check-task1").length).toBe(0);
		expect(screen.queryAllByTestId("check-task2").length).toBe(0);
		expect(screen.queryAllByTestId("check-task3").length).toBe(1);
	});
	test("check tasks", async () => {
		render(
			<Provider store={store}>
				<TasksList mode={"todo"} />
			</Provider>
		);
		expect(screen.queryAllByTestId("check-task1").length).toBe(1);
		expect(screen.queryAllByTestId("check-task2").length).toBe(1);
		userEvent.click(screen.getByTestId("check-task1"));
		await new Promise((r) => setTimeout(r, 2000));
		expect(screen.queryAllByTestId("check-task1").length).toBe(0);
		expect(screen.queryAllByTestId("check-task2").length).toBe(1);
	});
});
