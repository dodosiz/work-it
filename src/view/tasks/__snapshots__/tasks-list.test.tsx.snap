// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tasks list: done list 1`] = `
<div>
  <div
    class="task card"
  >
    <div
      class="card-body"
    >
      <div
        class="row"
      >
        <div
          class="col-md-1"
        >
          <div
            class="done"
            data-testid="check-task4"
          >
            <svg
              fill="currentColor"
              height="1.5em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 16 16"
              width="1.5em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M16 8A8 8 0 1 1 0 8a8 8 0 0 1 16 0zm-3.97-3.03a.75.75 0 0 0-1.08.022L7.477 9.417 5.384 7.323a.75.75 0 0 0-1.06 1.06L6.97 11.03a.75.75 0 0 0 1.079-.02l3.992-4.99a.75.75 0 0 0-.01-1.05z"
              />
            </svg>
          </div>
        </div>
        <div
          class="col-md-4"
        >
          <span
            class="pointer"
            data-testid="title-task4"
          >
            Develop the landing page.
          </span>
        </div>
        <div
          class="col-md-3"
        >
          John Doe
        </div>
        <div
          class="col-md-2"
        >
          Finished on: 2020-02-13
        </div>
        <div
          class="col-md-2"
        >
          <button
            class="btn btn-outline-danger"
            data-testid="delete-button-task4"
            type="button"
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 16 16"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M2.5 1a1 1 0 0 0-1 1v1a1 1 0 0 0 1 1H3v9a2 2 0 0 0 2 2h6a2 2 0 0 0 2-2V4h.5a1 1 0 0 0 1-1V2a1 1 0 0 0-1-1H10a1 1 0 0 0-1-1H7a1 1 0 0 0-1 1H2.5zm3 4a.5.5 0 0 1 .5.5v7a.5.5 0 0 1-1 0v-7a.5.5 0 0 1 .5-.5zM8 5a.5.5 0 0 1 .5.5v7a.5.5 0 0 1-1 0v-7A.5.5 0 0 1 8 5zm3 .5v7a.5.5 0 0 1-1 0v-7a.5.5 0 0 1 1 0z"
              />
            </svg>
          </button>
        </div>
      </div>
    </div>
  </div>
  <div
    class="task card"
  >
    <div
      class="card-body"
    >
      <div
        class="row"
      >
        <div
          class="col-md-1"
        >
          <div
            class="done"
            data-testid="check-task3"
          >
            <svg
              fill="currentColor"
              height="1.5em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 16 16"
              width="1.5em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M16 8A8 8 0 1 1 0 8a8 8 0 0 1 16 0zm-3.97-3.03a.75.75 0 0 0-1.08.022L7.477 9.417 5.384 7.323a.75.75 0 0 0-1.06 1.06L6.97 11.03a.75.75 0 0 0 1.079-.02l3.992-4.99a.75.75 0 0 0-.01-1.05z"
              />
            </svg>
          </div>
        </div>
        <div
          class="col-md-4"
        >
          <span
            class="pointer"
            data-testid="title-task3"
          >
            Fix the production bugs
          </span>
        </div>
        <div
          class="col-md-3"
        >
          John Doe
        </div>
        <div
          class="col-md-2"
        >
          Finished on: 2020-01-01
        </div>
        <div
          class="col-md-2"
        >
          <button
            class="btn btn-outline-danger"
            data-testid="delete-button-task3"
            type="button"
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 16 16"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M2.5 1a1 1 0 0 0-1 1v1a1 1 0 0 0 1 1H3v9a2 2 0 0 0 2 2h6a2 2 0 0 0 2-2V4h.5a1 1 0 0 0 1-1V2a1 1 0 0 0-1-1H10a1 1 0 0 0-1-1H7a1 1 0 0 0-1 1H2.5zm3 4a.5.5 0 0 1 .5.5v7a.5.5 0 0 1-1 0v-7a.5.5 0 0 1 .5-.5zM8 5a.5.5 0 0 1 .5.5v7a.5.5 0 0 1-1 0v-7A.5.5 0 0 1 8 5zm3 .5v7a.5.5 0 0 1-1 0v-7a.5.5 0 0 1 1 0z"
              />
            </svg>
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
